File: Assets/Scripts/Lua/NCSpeedLight/Core/Asset/AssetManager.lua
-----------------------------------------------
-- Copyright © 2014-2017 NCSpeedLight
--
-- FileName: AssetManager.lua
-- Describle:  资源管理器
-- Created By:  Wells Hsu
-- Date&Time:  2017/2/28 19:11:09
-- Modify History:
--
-----------------------------------------------
AssetManager =
{
	IsInitialited = false,
	Manifest = nil,
	Instance = nil,
	LoadedBundles = {},
	currentLoad = 0,
};

local this = AssetManager;

function AssetManager.Initialize()
	if AssetManager.IsInitialited == false then
		if SharedVariable.ASSETBUNDLE_MODE then
			Log.Info("AssetManager.Initialize");
			AssetManager.LoadAssetbundleManifest();
		end
	else
		Log.Error("AssetManager.Initialize: assetmanager has already been initialized.");
	end
end

function AssetManager.LoadAssetbundleManifest()
	local path = Constants.ASSET_BUNDLE_PATH .. Constants.ASSET_BUNDLE_MANIFEST_FILE;
	local bundle = UnityEngine.AssetBundle.LoadFromFile(path);
	if bundle ~= nil then
		AssetManager.Manifest = NCSpeedLight.Helper.LoadAssetFromBundle("AssetBundleManifest", typeof(UnityEngine.AssetBundleManifest), bundle);
		Log.Info("AssetManager.LoadAssetbundleManifest: load sccuss.");
	else
		Log.Error("AssetManager.LoadAssetbundleManifest: load error.");
	end
end

function AssetManager.LoadAsset(assetPath, type)
	if SharedVariable.ASSETBUNDLE_MODE then
		local index = string.find(assetPath, "/[^/]*$")
		local assetName = string.sub(assetPath, index + 1, string.len(assetPath));
		local bundleName = string.sub(assetPath, 1, index - 1);
		bundleName = string.gsub(bundleName, "/", "_");
		bundleName = string.lower(bundleName);
		bundleName = bundleName .. ".asset";
		local bundle = AssetManager.LoadAssetBundle(bundleName);
		if bundle ~= nil then
			return NCSpeedLight.Helper.LoadAssetFromBundle(assetName, type, bundle);
		else
			return nil;
		end
	else
		return UnityEngine.Resources.Load(assetPath, type);
	end
end

function AssetManager.LoadScene(sceneName)
	if SharedVariable.ASSETBUNDLE_MODE then
		local bundleName = "bundle_scenes.asset";
		local bundle = AssetManager.LoadAssetBundle(bundleName);
		if bundle == nil then
			Log.Error("AssetManager.LoadScene: can not load scene caused by nil scene bundle file.");
		else
			return UnityEngine.SceneManagement.SceneManager.LoadScene(sceneName);
		end
	else
		return UnityEngine.SceneManagement.SceneManager.LoadScene(sceneName);
	end
end

function AssetManager.UnloadScene(sceneName)
end

function AssetManager.LoadAssetBundle(bundleName)
	local bundleInfo = AssetManager.LoadedBundles[bundleName];
	if bundleInfo == nil or bundleInfo.Bundle == nil then
		AssetManager.LoadDependency(bundleName);
		local bundleFilePath = Constants.ASSET_BUNDLE_PATH .. bundleName;
		local bundle = UnityEngine.AssetBundle.LoadFromFile(bundleFilePath);
		bundleInfo = {RefCount = 1, Bundle = bundle};
		AssetManager.LoadedBundles[bundleName] = bundleInfo;
	else
		bundleInfo.RefCount = bundleInfo.RefCount + 1;
	end
	return bundleInfo.Bundle;
end

function AssetManager.LoadDependency(bundleName)
	local dependencies = AssetManager.Manifest:GetAllDependencies(bundleName);
	if dependencies ~= nil and dependencies.Length > 0 then
		for i = 0, dependencies.Length - 1 do
			local depName = dependencies[i];
			local bundleInfo = AssetManager.LoadedBundles[depName];
			if bundleInfo == nil or bundleInfo.Bundle == nil then
				local bundleFilePath = Constants.ASSET_BUNDLE_PATH .. depName;
				local bundle = UnityEngine.AssetBundle.LoadFromFile(bundleFilePath);
				bundleInfo = {RefCount = 1, Bundle = bundle};
				AssetManager.LoadedBundles[depName] = bundleInfo;
			else
				bundleInfo.RefCount = bundleInfo.RefCount + 1;
			end
		end
	end
end

function AssetManager.UnloadAssetBundle(assetPath)
	local index = string.find(assetPath, "/[^/]*$")
	local bundleName = string.sub(assetPath, 1, index - 1);
	bundleName = string.gsub(bundleName, "/", "_");
	bundleName = string.lower(bundleName);
	bundleName = bundleName .. ".asset";
	AssetManager.UnloadDependency(bundleName);
	local bundleInfo = AssetManager.LoadedBundles[bundleName];
	if bundleInfo ~= nil then
		bundleInfo.RefCount = bundleInfo.RefCount - 1;
		if bundleInfo.RefCount <= 0 then
			if bundleInfo.Bundle ~= nil then
				bundleInfo.Bundle:Unload(true);
				AssetManager.LoadedBundles[depName] = nil;
			end
		end
	end
end

function AssetManager.UnloadDependency(bundleName)
	local dependencies = AssetManager.Manifest:GetAllDependencies(bundleName);
	if dependencies ~= nil and dependencies.Length > 0 then
		for i = 0, dependencies.Length - 1 do
			local depName = dependencies[i];
			local bundleInfo = AssetManager.LoadedBundles[depName];
			if bundleInfo ~= nil then
				bundleInfo.RefCount = bundleInfo.RefCount - 1;
				if bundleInfo.RefCount <= 0 then
					if bundleInfo.Bundle ~= nil then
						bundleInfo.Bundle:Unload(true);
						AssetManager.LoadedBundles[depName] = nil;
					end
				end
			end
		end
	end
end

return this; 

File: Assets/Scripts/C#/NCSpeedLight/Core/Asset/AssetManager.cs
using System;
using System.Collections.Generic;
using UnityEngine;
using UnityEngine.SceneManagement;

namespace NCSpeedLight
{
    public static class AssetManager
    {

        public class BundleInfo
        {
            public AssetBundle Bundle;
            public int RefCount;
        }

        private static AssetBundleManifest m_Manifest;

        private static Dictionary<string, BundleInfo> m_LoadedBundles = new Dictionary<string, BundleInfo>();

        private static bool m_Isinit = false;

        public static void Initialize()
        {
            if (m_Isinit == false)
            {
                m_Isinit = true;
                if (Constants.ASSET_BUNDLE_MODE)
                {
                    LoadManifest();
                }
            }
        }

        private static void LoadManifest()
        {
            var path = Constants.LOCAL_ASSET_BUNDLE_PATH + Constants.ASSET_BUNDLE_MANIFEST_FILE;
            var bundle = AssetBundle.LoadFromFile(path);
            if (bundle != null)
            {
                m_Manifest = bundle.LoadAsset<AssetBundleManifest>("AssetBundleManifest");
            }
            else
            {
                Helper.LogError("AssetManager.LoadManifest: load error.");
            }
        }

        private static void LoadDependency(string bundleName)
        {
            string[] dependencies = m_Manifest.GetAllDependencies(bundleName);
            if (dependencies != null || dependencies.Length > 0)
            {
                for (int i = 0; i < dependencies.Length; i++)
                {
                    string depName = dependencies[i];
                    BundleInfo bundleInfo = null;
                    if (m_LoadedBundles.TryGetValue(depName, out bundleInfo) == false)
                    {
                        string bundleFilePath = Constants.LOCAL_ASSET_BUNDLE_PATH + depName;
                        AssetBundle bundle = AssetBundle.LoadFromFile(bundleFilePath);
                        bundleInfo = new BundleInfo() { Bundle = bundle, RefCount = 1 };
                        if (m_LoadedBundles.ContainsKey(depName))
                        {
                            m_LoadedBundles.Remove(depName);
                        }
                        m_LoadedBundles.Add(depName, bundleInfo);
                    }
                    else
                    {
                        bundleInfo.RefCount++;
                    }
                }
            }
        }

        private static AssetBundle LoadAssetBundle(string bundleName)
        {
            BundleInfo bundleInfo = null;
            if (m_LoadedBundles.TryGetValue(bundleName, out bundleInfo) == false)
            {
                LoadDependency(bundleName);
                string bundleFilePath = Constants.LOCAL_ASSET_BUNDLE_PATH + bundleName;
                AssetBundle bundle = AssetBundle.LoadFromFile(bundleFilePath);
                bundleInfo = new BundleInfo()
                {
                    RefCount = 1,
                    Bundle = bundle
                };
                if (m_LoadedBundles.ContainsKey(bundleName))
                {
                    m_LoadedBundles.Remove(bundleName);
                }
                m_LoadedBundles.Add(bundleName, bundleInfo);
                return bundleInfo.Bundle;
            }
            else
            {
                bundleInfo.RefCount = bundleInfo.RefCount + 1;
                return bundleInfo.Bundle;
            }
        }

        private static void UnloadDependency(string bundleName)
        {
            string[] dependencies = m_Manifest.GetAllDependencies(bundleName);
            if (dependencies != null || dependencies.Length > 0)
            {
                for (int i = 0; i < dependencies.Length; i++)
                {
                    string depName = dependencies[i];
                    BundleInfo bundleInfo = null;
                    if (m_LoadedBundles.TryGetValue(depName, out bundleInfo))
                    {
                        bundleInfo.RefCount--;
                        if (bundleInfo.Bundle == null)
                        {
                            m_LoadedBundles.Remove(depName);
                        }
                        else if (bundleInfo.RefCount <= 0)
                        {
                            bundleInfo.Bundle.Unload(true);
                            m_LoadedBundles.Remove(depName);
                        }
                    }
                }
            }
        }

        private static void UnloadAssetBundle(string bundleName)
        {
            BundleInfo bundleInfo = null;
            if (m_LoadedBundles.TryGetValue(bundleName, out bundleInfo))
            {
                UnloadDependency(bundleName);
                bundleInfo.RefCount--;
                if (bundleInfo.Bundle == null)
                {
                    m_LoadedBundles.Remove(bundleName);
                }
                else if (bundleInfo.RefCount <= 0)
                {
                    bundleInfo.Bundle.Unload(true);
                    m_LoadedBundles.Remove(bundleName);
                }
            }
        }
       
        public static UnityEngine.Object LoadAsset(string assetPath, Type type)
        {
            if (Constants.ASSET_BUNDLE_MODE)
            {
                int index = assetPath.LastIndexOf("/");
                string assetName = assetPath.Substring(index + 1);
                string bundleName = assetPath.Substring(0, index);
                bundleName = bundleName.Replace("/", "_");
                bundleName = bundleName.ToLower();
                bundleName = bundleName + Constants.ASSET_BUNDLE_FILE_EXTENSION;
                AssetBundle bundle = LoadAssetBundle(bundleName);
                if (bundle != null)
                {
                    return bundle.LoadAsset(assetName, type);
                }
                else
                {
                    return null;
                }
            }
            else
            {
                return Resources.Load(assetPath, type);
            }
        }

        public static void UnloadAsset(string assetPath)
        {
            if (Constants.ASSET_BUNDLE_MODE)
            {
                int index = assetPath.LastIndexOf("/");
                string assetName = assetPath.Substring(index + 1);
                string bundleName = assetPath.Substring(0, index);
                bundleName = bundleName.Replace("/", "_");
                bundleName = bundleName.ToLower();
                bundleName = bundleName + Constants.ASSET_BUNDLE_FILE_EXTENSION;
                UnloadAssetBundle(bundleName);
            }
        }

        public static void LoadScene(string sceneName)
        {
            if (Constants.ASSET_BUNDLE_MODE)
            {
                string bundleName = "bundle_scenes.asset";
                AssetBundle bundle = LoadAssetBundle(bundleName);
                if (bundle == null)
                {
                    Helper.LogError("AssetManager.LoadScene: can not load scene caused by nil scene bundle file.");
                }
                else
                {
                    SceneManager.LoadScene(sceneName);
                }
            }
            else
            {
                SceneManager.LoadScene(sceneName);
            }
        }

        public static void UnloadScene(string sceneName)
        {
            //UnloadDependency("bundle_scenes.asset");
        }

    }
}